pkg/kv/kvserver/multiqueue/multi_queue.go

```diff
@@ -0,0 +1,231 @@
+			task.permitC <- &Permit{valid: true}
+			m.remainingRuns--
+			m.lastQueueIndex = index
+			return
+		}
+	}
+}
+
+// Add returns a Task that must be closed (calling Task.Close) to
+// release the Permit. The number of types is expected to
+// be relatively small and not be changing over time.
+func (m *MultiQueue) Add(queueType int, priority float64) *Task {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// The mutex starts locked, unlock it when we are ready to run.
+	pos, ok := m.mapping[queueType]
+	if !ok {
+		// Append a new entry to both mapping and outstanding each time there is
+		// a new queue type.
+		pos = len(m.outstanding)
+		m.mapping[queueType] = pos
+		m.outstanding = append(m.outstanding, notifyHeap{})
+	}
+	newTask := Task{
+		priority:  priority,
+		permitC:   make(chan *Permit, 1),
+		heapIdx:   -1,
+		queueType: queueType,
+	}
+	heap.Push(&m.outstanding[pos], &newTask)
+
+	// Once we are done adding a task, signal the main loop in case it finished
+	// all its work and was waiting for more work. We are holding the mu lock when
+	// signaling, so we guarantee that it will not be able to respond to the
+	// signal until after we release the lock.
+	m.tryRunNextLocked()
+
+	return &newTask
+}
+
+// Cancel will cancel a Task that may not have started yet. This is useful if it
+// is determined that it is no longer required to run this Task.
+func (m *MultiQueue) Cancel(task *Task) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// Find the right queue and try to remove it. Queues monotonically grow, and a
+	// Task will track its position within the queue.
+	queueIdx := m.mapping[task.queueType]
+	ok := m.outstanding[queueIdx].tryRemove(task)
+	// If we get here, we are racing with the task being started. The concern is
+	// that the caller may also call MultiQueue.Release since the task was
+	// started. Either we get the permit or the caller, so we guarantee only one
+	// release will be called.
+	if !ok {
+		select {
+		case p, ok := <-task.permitC:
+			// Only release if the channel is open, and we can get the permit.
+			if ok {
+				p.valid = false
+				m.remainingRuns++
+				m.tryRunNextLocked()
+			}
+		default:
+			// If we are not able to get the permit, this means the permit has already
+			// been given to the caller, and they must call Release on it.
+		}
+	}
+}
+
+// Release needs to be called once the Task that was running has completed and
+// is no longer using system resources. This allows the MultiQueue to call the
+// next Task.
+func (m *MultiQueue) Release(permit *Permit) {
+	if !permit.valid {
+		panic("double release of permit")
+	}
+	permit.valid = false
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// We released one, so we can run one more now.
+	m.remainingRuns++
+	m.tryRunNextLocked()
+}
+
+// Len returns the number of additional tasks that can be added without
+// queueing. This will return 0 if there is anything queued. This method should
+// only be used for testing.
```

Comment:
If this method is only for testing then you can actually define it in `multi_queue_test.go`, because that's in the same package.